src/spi.rs

```diff
@@ -0,0 +1,222 @@
+    expected_mode: Mode,
+    expected_data: Vec<u8>,
+    response: Vec<u8>,
+}
+
+impl Transaction {
+    /// Create a write transaction
+    pub fn write(expected: Vec<u8>) -> Transaction {
+        Transaction {
+            expected_mode: Mode::Write,
+            expected_data: expected,
+            response: Vec::new(),
+        }
+    }
+
+    /// Create a transfer transaction
+    pub fn transfer(expected: Vec<u8>, response: Vec<u8>) -> Transaction {
+        Transaction {
+            expected_mode: Mode::Transfer,
+            expected_data: expected,
+            response,
+        }
+    }
+}
+
+/// Mock SPI implementation
+/// This supports the specification and checking of expectations to allow automated testing of SPI based drivers.
+/// Mismatches between expected and real SPI transactions will cause runtime assertions to assist with locating faults.
+pub struct SPIMock {
+    expected: Arc<Mutex<VecDeque<Transaction>>>,
+}
+
+impl SPIMock {
+    /// Create a new mock SPI interface
+    pub fn new() -> Self {
+        SPIMock {
+            expected: Arc::new(Mutex::new(VecDeque::new())),
+        }
+    }
+
+    /// Set expectations on the SPI interface
+    /// This is a list of SPI transactions to be executed in order
+    /// Note that setting this will overwrite any existing expectations
+    pub fn expect(&mut self, expected: Vec<Transaction>) {
+        let mut e = self.expected.lock().unwrap();
+        *e = expected.into();
+    }
+
+    /// Assert that all expectations on a given SPIMock have been met
+    pub fn done(&mut self) {
+        let expected = self.expected.lock().unwrap();
+        assert_eq!(expected.len(), 0);
+    }
+}
+
+impl spi::Write<u8> for SPIMock {
+    type Error = MockError;
+
+    /// spi::Write implementation for SPIMock
+    /// This will cause an assertion if the write call does not match the next expectation
+    fn write(&mut self, buffer: &[u8]) -> Result<(), Self::Error> {
+        let w = self
+            .expected
+            .lock()
+            .unwrap()
+            .pop_front()
+            .expect("no expectation for spi::write call");
+        assert_eq!(w.expected_mode, Mode::Write);
+        assert_eq!(&w.expected_data, &buffer);
+        Ok(())
+    }
+}
+
+impl spi::Transfer<u8> for SPIMock {
+    type Error = MockError;
+
+    /// spi::Transfer implementation for SPIMock
+    /// This writes the provided response to the buffer and will cause an assertion if the written data does not match the next expectation
+    fn transfer<'w>(&mut self, buffer: &'w mut [u8]) -> Result<&'w [u8], Self::Error> {
+        let w = self
+            .expected
+            .lock()
+            .unwrap()
+            .pop_front()
+            .expect("no expectation for spi::transfer call");
+        assert_eq!(w.expected_mode, Mode::Transfer);
+        assert_eq!(&w.expected_data, &buffer);
+        for i in 0..w.response.len() {
+            buffer[i] = w.response[i];
```

Comment:
How do you know that the buffer has the same size as the response? This isn't guaranteed by the SPI trait, right?

Furthermore, is it expected that the response bytes are written to the buffer? It seems that in the [trait](https://docs.rs/embedded-hal/0.2.1/embedded_hal/blocking/spi/trait.Transfer.html#tymethod.transfer) the return reference lifetime is the same as the input buffer lifetime, so that's probably the only way for an SPI implementation to work? (Unfortunately the trait docs don't say anything about that...)

Reply:
Fair point, it is inherent to the way SPI works, so I have added a length check to the transfer function.

It is expected / is the only way it works afaik.